### PR TITLE
refactor: Avoid implicit reexport in submodules

### PR DESCRIPTION
**Pull request by l0b0**
**Related Issue(s):** #332

**Description:** I couldn't do the equivalent changes anywhere in the `serialization` submodule without running into circular imports, so I've postponed that until it's easier to fix.

**PR Checklist:**

- [x] Code is formatted (run `pre-commit run --all-files`)
- [x] Tests pass (run `scripts/test`)
- [ ] Documentation has been updated to reflect changes, if applicable
- [x] This PR maintains or improves overall codebase code coverage.
- [ ] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/develop/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

**Comment by duckontheweb** [comment]
Sorry, I just realized I merged some changes from a more recent PR that created conflicts. I'm happy to rebase and resolve these if you want.

**Review by duckontheweb — COMMENTED** [review]
I'm +1 for this change, but I'd like to be sure others are on board as well.

File: pystac/extensions/file.py [hunk 1]
```diff
@@ -6,7 +6,10 @@
 from enum import Enum
 from typing import Any, Dict, List, Optional
 
-import pystac
+from pystac.asset import Asset
+from pystac.stac_object import STACObjectType
+from pystac.errors import ExtensionTypeError
+from pystac.item import Item
 from pystac.extensions.base import ExtensionManagementMixin, PropertiesExtension
```
Comment (schwehr) [inline pystac/extensions/file.py:13]: Here I would suggest: `from pystac.extentions import base`.  Then it's `base.PropertiesExtension` in the code.

**Review by schwehr — COMMENTED** [review]
Some groups have style guides and other reasons to follow specific patterns.  The one I'm suggesting is documented here: https://google.github.io/styleguide/pyguide.html#22-imports

**Comment by duckontheweb** [comment]
> The one I'm suggesting is documented here: https://google.github.io/styleguide/pyguide.html#22-imports

I like this approach for the reasons that @schwehr mentions in [this comment](https://github.com/stac-utils/pystac/pull/539#discussion_r670593851). It also keeps our import statements a bit shorter and means that classes/functions with the same name in different modules wouldn't conflict. We would have to resolve some shadowing issues since we use variable names like `link` in the code and these would shadow imports like `from pystac import link`. 

If we were to adopt this, I would also propose that in cases where we are importing a class purely for type annotations we always put it in a `if TYPE_CHECKING:` block and import the class directly for consistency with the way `typing` and `typing_extensions` types are imported.

File: pystac/extensions/base.py [hunk 2]
```diff
@@ -1,7 +1,12 @@
 from abc import ABC, abstractmethod
 from typing import Generic, Iterable, List, Optional, Dict, Any, Type, TypeVar, Union
 
-import pystac
+from pystac.asset import Asset
+from pystac.summaries import Summaries
+from pystac.stac_object import STACObject
+from pystac.collection import Collection
+from pystac.errors import ExtensionNotImplemented
+from pystac.summaries import RangeSummary
```
[the remark this answers is not in the document]
Comment (l0b0) [reply pystac/extensions/base.py:9]: That's a nice compromise. My main question would be whether it helps with avoiding circular imports, which are currently plaguing the code base. For example, running `isort` will immediately break the entire code base, and there seems to be a bunch of duplicate imports (`import pystac` followed by `from pystac import foo`) which, if merged, also cause circular import errors.

File: pystac/extensions/base.py [hunk 3]
```diff
@@ -12,16 +17,16 @@ class SummariesExtension:
     extension-specific class that inherits from this class and instantiate that. See
     :class:`~pystac.extensions.eo.SummariesEOExtension` for an example."""
 
-    summaries: pystac.Summaries
+    summaries: Summaries
```
[the remark this answers is not in the document]
Comment (l0b0) [reply pystac/extensions/base.py:20]: Yeah, fixing the comments is another can of worms I'm not comfortable with tackling until we've nailed the imports themselves. I'm guessing they should resolve within the current set of imports, that is, if I `from pystac import summaries` the comment should be `summaries: summaries.Summaries` as you said.

**Comment by l0b0** [comment]
> Sorry, I just realized I merged some changes from a more recent PR that created conflicts. I'm happy to rebase and resolve these if you want.

Thanks, that would be helpful. Then I can start changing all these PRs to use the import strategy suggested by @schwehr.

**Comment by l0b0** [comment]
I'm going to redo these PRs from scratch, so this may take a while…

**Comment by duckontheweb** [comment]
> > Sorry, I just realized I merged some changes from a more recent PR that created conflicts. I'm happy to rebase and resolve these if you want.
> 
> Thanks, that would be helpful. Then I can start changing all these PRs to use the import strategy suggested by @schwehr.

@l0b0 Not sure if this is still needed if you are starting from scratch, but I rebased this onto `main`.

**Comment by lossyrob** [comment]
I'm -0 on this change. I really like the simplicity of just using `import pystac` inside the library and going from there. I'm confused about what we buy by making the import list more verbose.

That said, I don't feel that strongly and if people think it will really improve the maintainability in the long run then I'm ok with these sort of changes for the sake of strictness.

I do feel strongly that, as a user, I don't want to have to go digging for imports to use the types of PySTAC. Having a single `import pystac` and being able to work from there without having to worry about more type or submodule imports is a good developer experience, and I'd like to maintain that ability (which may be more of a response to https://github.com/stac-utils/pystac/pull/543#issuecomment-880810202). This will be especially true of users who aren't using the same tooling as someone with a more full development environment - think about the data scientists using notebooks that don't have the same autoimport functionality we might be working with as library developers, but who are and will be (I believe) an increasing percentage of the user base. This is already something I've received some negative feedback on with the removal of the magic `.ext` method and requiring extensions be verbosely imported instead, which is a change I made for more type-aware coding (and I'll stand by that, as the `.ext` method would never be able to even give tab completion, errored in strange ways and was hard to debug).

I'm not particularly in favor of the "only import submodules rule", but that's just not the way I've coded Python. Again, if that's what folks agree is a good way to go for the internal library development, then I'm willing to go with it - it feels like preference tuning at this point which I guess marks a good point in the library's lifecycle if we're getting to that fine of detail. I'd just keep the developer experience of downstream users who don't have strongly typed tooling in mind when making changes that will affect the import strategy for those users (which may not apply to this PR but is more a general comment).

**Pull request closed without merging**